Rewriting the code for easier implementation of new features

SimonBrandner opened the pull request:
I have created a `Workspace` class to store the workspace information. Then, I have added a static method `Workspace::filterByCurrentScreen()` to filter workspaces by the current screen. Then I rewrote the already existing code to use the `Workspace` class and the `Workspace::filterByCurrentScreen()` method.

This should simplify implementing new features such as showing all workspaces instead of only those on the current screen.

I have also added clang-format.

duvholt commented on src/i3pager.h:
```diff
@@ -33,5 +41,4 @@ public Q_SLOTS:
 
-
-#endif //I3PAGER_PLASMOID_H
+#endif I3PAGER_PLASMOID_H
```
When compiling I'm getting some warnings about these endif changes:
```
i3-pager/src/i3pager.h:44:8: warning: extra tokens at end of #endif directive [-Wendif-labels]
   44 | #endif I3PAGER_PLASMOID_H
```
I don't know C++ that well, but from some quick Googling it seems like endif with a comment is more correct?

SimonBrandner replied:
I don't know C++ that well either.. You're absolutely right, what I wrote doesn't make any sense. I'll fix it.

duvholt commented on src/i3pager.h:
```diff
@@ -1,26 +1,34 @@
     Q_OBJECT
     Q_PROPERTY(QString currentScreen WRITE setCurrentScreen)
-    Q_PROPERTY(QVariantList workspaces READ getWorkspaces NOTIFY currentScreenChanged)
+    Q_PROPERTY(QList<Workspace> workspaces READ getWorkspaces NOTIFY currentScreenChanged)
```
Much better with some actual types than the QVariantList approach :+1:

duvholt reviewed (approved):
Thanks for cleaning up the codebase and adding clang format.

Is this ready to be merged? I saw you added some new commits today.

SimonBrandner commented:
> Is this ready to be merged? I saw you added some new commits today.

You're welcome. I originally thought it was ready to be merged but as you've said I've made some changes recently and I'll probably do some more, so I'd rather wait.

SimonBrandner commented:
I think I am done with the refactoring. This is ready to be merged if you agree. 

To sum up today's changes: 

+ Added `focused` property to the `Workspace` class.
+ Renamed `currentScreenChanged()` to `workspacesChanged()`. 
+ I fixed the include guards and some other warnings.
+ Rewrote the code so that `repeater.model` is set by a function. 
+ Added a `filterByCurrentScreen` parameter to `I3Pager::getWorkspaces()`.

The main focus was to allow retrieving the workspaces using `I3Pager::getWorkspaces()` method. Parameters can now be added to it. Based on these parameters workspaces could be filtred and or ordered.

duvholt commented:
I was about to merge this, but then I noticed a bug. After the workspaces Q_PROPERTY was removed the plasmoid stopped working in plasmashell. It still works with plasmoidviewer so I don't really get why it stopped working.
Based on the Qt documentation it doesn't seem like it should be necessary to use a property, but I can't get it to work in plasmashell without adding back the property. 

I did a bisect and this is the commit that broke the plasmoid for me: https://github.com/duvholt/i3-pager/pull/6/commits/9dd07431c7108d084f776f905ceb8a3ffdca4e19

duvholt commented:
Okay, did some more digging and this is likely just an issue on my system. Doesn't seem like the qml part of the plasmoid is updated on install anymore.

duvholt commented:
Well, this is silly. Looks like I had managed to install a copy of the plasmoid in `~/.local/share/plasma/plasmoids` which had priority over `/usr/share/plasma/plasmoids`. Everything is working now.
